bookserver/crud.py

```diff
@@ -494,3 +580,189 @@ async def create_user_chapter_progress_entry(
     async with async_session.begin() as session:
         session.add(new_ucp)
     return UserChapterProgressValidator.from_orm(new_ucp)
+
+
+#
+# Select Question Support
+# -----------------------
+
+
+async def create_selected_question(
+    sid: str,
+    selector_id: str,
+    selected_id: str,
+    points: Optional[int] = None,
+    competency: Optional[str] = None,
+) -> SelectedQuestionValidator:
+    new_sqv = SelectedQuestion(
+        sid=sid,
+        selector_id=selector_id,
+        selected_id=selected_id,
+        points=points,
+        competency=competency,
+    )
+    async with async_session.begin() as session:
+        session.add(new_sqv)
+    return SelectedQuestionValidator.from_orm(new_sqv)
+
+
+async def fetch_selected_question(sid: str, selector_id) -> SelectedQuestionValidator:
+    query = select(SelectedQuestion).where(
+        (SelectedQuestion.sid == sid) & (SelectedQuestion.selector_id == selector_id)
+    )
+
+    async with async_session() as session:
+        res = await session.execute(query)
+        rslogger.debug(f"{res=}")
+        return SelectedQuestionValidator.from_orm(res.scalars().first())
+
+
+async def update_selected_question(sid: str, selector_id: str, selected_id: str):
+    stmt = (
+        update(SelectedQuestion)
+        .where(
+            (SelectedQuestion.sid == sid)
+            & (SelectedQuestion.selector_id == selector_id)
+        )
+        .values(selected_id=selected_id)
+    )
+    async with async_session.begin() as session:
+        await session.execute(stmt)
+    rslogger.debug("SUCCESS")
+
+
+# Questions and Assignments
+# -------------------------
+
+
+async def fetch_question(
+    name: str, basecourse: Optional[str] = None
+) -> QuestionValidator:
+
+    where_clause = Question.name == name
+    if basecourse:
+        where_clause = where_clause & (Question.base_course == basecourse)
+
+    query = select(Question).where(where_clause)
+
+    async with async_session() as session:
+        res = await session.execute(query)
+        rslogger.debug(f"{res=}")
+        return QuestionValidator.from_orm(res.scalars().first())
+
+
+auto_gradable_q = [
+    "clickablearea",
+    "mchoice",
+    "parsonsprob",
+    "dragndrop",
+    "fillintheblank",
+]
+
+
+async def fetch_matching_questions(request_data: schemas.SelectQRequest) -> List[str]:
+
+    if request_data.questions:
+        questionlist = request_data.questions.split(",")
```

Comment:
Aack, no JSON encoding for the data sent? Seems fragile.

Reply:
I don't understand the comment.  request_data is a schema validator object.  what more do you want?

Reply:
I mean the JS that sends the request data would ideally encode it as JSON instead of joining values with a comma. This is more a note for what needs fixing in the future.